File: Regression/4. Polynomial Regression/polynomial_regression.py
# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# We always want to consider matrix of features as a matrix and not as a vector
# Here python will not consider the upper bound index 2 in X
# (10,1) -> Matrix of feature of 10 lines and 1 column
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Here we need the shole data as training set to predict the accurate predictions
# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# We dont need Feature Scaling because the polynomial regression means adding some polynimials into multiple linear regression equation
# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting linear regression to the dataset
# Here 2nd polynomial regressor is based on linear regressor on which we add some polynomial terms
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

# Fitting polynomial regression to the dataset
# Here we include a new class to include polynomial terms into linear regression equation -> polynomial features class
from sklearn.preprocessing import PolynomialFeatures
# Here poly_reg object is an transformation tool that will transform matrix of features X into a new matrix of features Xpoly (creates new independent varibles according to the formula X^2, X^3 ..... X^n)
# We call the PolynomialFeatures class with the parameter degree.
# degree -> Specifies the degree of polynomial features in the future feature scale matrixs Xpoly
poly_reg = PolynomialFeatures(degree = 3)
# As degree = 2 we are adding only 1 additional polynomial term
# Creating Xpoly matrix  
X_poly = poly_reg.fit_transform(X)
# The pupose of 2nd linear regressor lin_reg2 is to include this fit made with poly_reg object into our regression model
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y) 

# Visualizing the Linear Regression Results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Level of Position ()')
plt.ylabel('Salaries')
plt.show()

# Visualizing the Polynomial Regression Results
# To get a curve with a higher resolution
# X_grid contains all the levels plus incremental steps between the levels with a resolution of 0.1
# mix(X) -> lower bound
# max(X) -> upper bound
# Incrementation = 0.1
X_grid = np.arange(min(X), max(X), 0.1) # This gives a vector
# To reshap vector into a matrix
# We reshape X_grid into a matrix where the number of lines is the number of elements of X_grid [len(X_grid] and number of columns is 1
X_grid = X_grid.reshape((len(X_grid)), 1)
plt.scatter(X, y, color = 'red')
# Our model predicts imaginary salaries of 90 levels from 1 to 10 with a resolution of 0.1
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Level of Position ()')
plt.ylabel('Salaries')
plt.show()

# Adding a degree to Polynomial Regression Model
# Make degree = 3
# Here we can see that it is not convex anymore
# Our model is much better now

# Adding another degree
# Making degree = 4 we get a perfect model

# Predicting a new result with Linear Regression
lin_reg.predict(6.5)

# Predicting a new result with Polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Regression/4. Polynomial Regression/polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# Here we wont need feature scaling as a polynomial regression model is actually a multiple rgression model with polynomial terms 
# Instead of having different features like features that represent something very different
# Here we are taking the first feature that is the position levels form 1 to 10 and as in the case of multiple line regression we take the squares of the levels here
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Fitting Polynomial Regression to the dataset
# Here the independent variables are actually the polynomial features and here we have only one independent variable
# Adding polynomial features  -> These are some aditional independent variables that are Level^2, Level^3 and so on
# These compose our new matrix of features in some way which will be the matrix on which we apply our multiple linear regression
# This will make the whole model into Polynomial Regression Model
# In short a polynomial regression model is a multiple linear regression that composed of one independent variable and additional independent variables that are the polynomial terms of the first independent variables
# Adding a new independent variable Level 2 (Level^2). To add a column in a dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
# Level -> orginal independent Variables
# Level2 -> Polynomial Terms of first independent variables (Level)
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualizing the Linear Regression Results
#install.packages('ggplot2')
library(ggplot2)
ggplot() + geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), colour = 'blue') +
              ggtitle('Truth or Bluff (Linear Regression)') +
              xlab('Level') +
              ylab('Salaries')

# Visualizing the Polynomial Regression Results
ggplot() + geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)), colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salaries')

# Predicting a new Result with Linear Regression Model
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predicting a new Result with Polynomial Regression Model
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                                      Level2 = 6.5^2,
                                      Level3 = 6.5^3,
                                      Level4 = 6.5^4))
